lib/std/SemanticVersion.zig

```diff
@@ -0,0 +1,295 @@
+        "1.0.0-alpha_beta",
+        "-alpha.",
+        "1.0.0-alpha..",
+        "1.0.0-alpha..1",
+        "1.0.0-alpha...1",
+        "1.0.0-alpha....1",
+        "1.0.0-alpha.....1",
+        "1.0.0-alpha......1",
+        "1.0.0-alpha.......1",
+        "01.1.1",
+        "1.01.1",
+        "1.1.01",
+        "1.2",
+        "1.2.3.DEV",
+        "1.2-SNAPSHOT",
+        "1.2.31.2.3----RC-SNAPSHOT.12.09.1--..12+788",
+        "1.2-RC-SNAPSHOT",
+        "-1.0.3-gamma+b7718",
+        "+justmeta",
+        "9.8.7+meta+meta",
+        "9.8.7-whatever+meta+meta",
+    }) |invalid| expectError(error.InvalidVersion, parse(invalid));
+
+    // Valid version string that may overflow.
+    const big_valid = "99999999999999999999999.999999999999999999.99999999999999999";
+    if (parse(big_valid)) |ver| try testFmt(big_valid, "{}", .{ver}) else |err| expect(err == error.Overflow);
+
+    // Invalid version string that may overflow.
+    const big_invalid = "99999999999999999999999.999999999999999999.99999999999999999----RC-SNAPSHOT.12.09.1--------------------------------..12";
+    if (parse(big_invalid)) |ver| std.debug.panic("expected error, found {}", .{ver}) else |err| {}
+}
+
+test "SemanticVersion precedence" {
+    // SemVer 2 spec 11.2 example: 1.0.0 < 2.0.0 < 2.1.0 < 2.1.1.
+    expect(order(try parse("1.0.0"), try parse("2.0.0")) == .lt);
+    expect(order(try parse("2.0.0"), try parse("2.1.0")) == .lt);
+    expect(order(try parse("2.1.0"), try parse("2.1.1")) == .lt);
+
+    // SemVer 2 spec 11.3 example: 1.0.0-alpha < 1.0.0.
+    expect(order(try parse("1.0.0-alpha"), try parse("1.0.0")) == .lt);
+
+    // SemVer 2 spec 11.4 example: 1.0.0-alpha < 1.0.0-alpha.1 < 1.0.0-alpha.beta < 1.0.0-beta <
+    // 1.0.0-beta.2 < 1.0.0-beta.11 < 1.0.0-rc.1 < 1.0.0.
+    expect(order(try parse("1.0.0-alpha"), try parse("1.0.0-alpha.1")) == .lt);
+    expect(order(try parse("1.0.0-alpha.1"), try parse("1.0.0-alpha.beta")) == .lt);
+    expect(order(try parse("1.0.0-alpha.beta"), try parse("1.0.0-beta")) == .lt);
+    expect(order(try parse("1.0.0-beta"), try parse("1.0.0-beta.2")) == .lt);
+    expect(order(try parse("1.0.0-beta.2"), try parse("1.0.0-beta.11")) == .lt);
+    expect(order(try parse("1.0.0-beta.11"), try parse("1.0.0-rc.1")) == .lt);
+    expect(order(try parse("1.0.0-rc.1"), try parse("1.0.0")) == .lt);
+}
+
+// This is copy-pasted from fmt.zig since it is not public.
+fn testFmt(expected: []const u8, comptime template: []const u8, args: anytype) !void {
+    var buf: [100]u8 = undefined;
+    const result = try std.fmt.bufPrint(buf[0..], template, args);
+    if (std.mem.eql(u8, result, expected)) return;
+
+    std.debug.warn("\n====== expected this output: =========\n", .{});
+    std.debug.warn("{}", .{expected});
+    std.debug.warn("\n======== instead found this: =========\n", .{});
+    std.debug.warn("{}", .{result});
+    std.debug.warn("\n======================================\n", .{});
+    return error.TestFailed;
+}
```

Comment:
What if we move it into `std.testing` and make it public? Would that make sense?

Reply:
Yes, that makes sense to me. I suppose I should do that in a separate PR.

Reply:
PR created: https://github.com/ziglang/zig/pull/6831